### ISSUE TITLE
[BUG] Not working with new library beta

### ISSUE THREAD

**Issue by wickles** [Depressurizer/Depressurizer#165]
Using depressurizer version 4.12.1, I removed and edited some categories and export the config with steam closed, then open steam and my edits do not appear to have been applied (no change in collections area or on game pages). I'd guess they are storing the information differently now?

**Comment by Zhuinden** [comment]
The new library beta completely eliminated all of my previously added categories, both automated and manual. I hope Steam will do something about it (although I have my doubts).

**Comment by neptuwunium** [comment]
It didn't for me, and as a result I have a few hundred categories lagging Steam to run at 2 FPS.

**Comment by matrixraven** [comment]
I organized my games a few weeks back with the pre-release version, and I updated to the steam beta library last night. Everything looks great. I have both the tags that depressurizer made, as well as my previous categories that I manually made.
![2019-09-18 08_51_53-Steam](https://user-images.githubusercontent.com/31789125/65164437-bd6e9f00-d9f1-11e9-9673-a46db3ed3cfa.jpg)

**Comment by Zhuinden** [comment]
I wish I knew if I can still restore my categories... I guess I'll try to roll back, do a Depressurize, then update, then hope for the best. 

**EDIT:** Rollbacking restores my categories... for now.

**Comment by wickles** [comment]
Are you sure they're actually gone? You have to click on the collections button to see them now. They also show up on the top right of the game page, over the banner.

**Comment by Zhuinden** [comment]
Well it says `Uncategorized (2358)` so that looks uncategorized to me.

**Comment by ltctceplrm** [comment]
> 
> 
> Well it says `Uncategorized (2358)` so that looks uncategorized to me.

Same problem here, lost all categories but if I roll back I have them all back again, its just in the beta that they all are uncategorized.

**Comment by seanniu93** [comment]
I think the beta collections and original categories are separate now. Either when the beta rolled out or you enrolled in the beta, the collections forked the categories (at least it did for me), but any changes made to either beta or original after the fork don't sync.

@healingbrew try turning on hardware acceleration or the low performance option

**Comment by caenissnow** [comment]
I hadn't run Depressurizer in a while and i had gotten more games for my collection.  I was excited to see how it worked with the new Steam library and so I loaded up Depressurizer to update all the recent games and while it looks like everything updated fine in Depressurizer, Steam is not updating with the newer profile.  For instance, I am seeing 210 Hidden Object games in Depressurizer but still only seeing 175 in Steam, the amount I had in that tag before the update.  Is there something i'm doing wrong?  Thanks.

**Comment by thatoekyaw** [comment]
Anyone know where the collections are saved on the beta now? used to be sharedconfig.vdf on the 'old' library.

**Comment by neptuwunium** [comment]
It's stored in Steam's internal CEF browser's localstorage (aka LevelDB) at `%LOCALAPPDATA%\Steam\htmlcache\Local Storage\leveldb`

The key in leveldb is the standard local stoarge key (i.e. _host\0\u0001key), which in this case is `_https://steamloopback.host\0\u0001U[steamid]-cloud-storage-namespace-1`

The value for this is a UTF-16-BE encoded JSON array (with no BOM.) Steam keeps deleted collections.

```json
["user-collections.uc-somerandomstring", {
    "key": "user-collections.uc-thesamerandomstring",
    "timestamp": 1568808979,
    "value": "{\"id\":\"uc-thesamerandomstring\",\"name\":\"unplayed\",\"added\":[],\"removed\":[],\"filterSpec\":{\"nFormatVersion\":2,\"strSearchText\":\"\",\"filterGroups\":[{\"rgOptions\":[],\"bAcceptUnion\":true},{\"rgOptions\":[4],\"bAcceptUnion\":false},{\"rgOptions\":[],\"bAcceptUnion\":false},{\"rgOptions\":[],\"bAcceptUnion\":false},{\"rgOptions\":[],\"bAcceptUnion\":false}]}}"
}]
```

Favourites looks like

```json
    ["user-collections.favorite", {
        "key": "user-collections.favorite",
        "timestamp": 1568881294,
        "value": "{\"id\":\"favorite\",\"name\":\"Favorites\",\"added\":[502500],\"removed\":[260130]}"
    }],
```

Proof of Concept: https://github.com/healingbrew/SteamCollectionsDumper/blob/master/Program.cs
This will export all cloud-storage keys in local storage to the database folder, or import them if they're already exported.

So TL;DR - it's extremely hard for someone to just go in and edit the collections. You have to actively manipulate local storage objects.

Creating normal collections (as in category/tag-like) works, but it wipes dynamic collections. This might be because the random string isn't actually random or there's some encoding somewhere that's lost and breaks the dataset. In testing this I accidentally wiped my steam collections/favorites several times, but it restored fine with the initial backups i made.

**Comment by thefunkygibbon** [comment]
So what does this mean exactly?  is it something that can be worked around or does it effectively make depressurizer useless?   
I too am (obviously) seeing this issue, although with mine it is just not doing any changes, it hasn't so far corrupted my steam groups or deleted any.

**Comment by neptuwunium** [comment]
You can work around it, but it's a lot more... unpredictable.

**Comment by ltctceplrm** [comment]
I opened a ticket with steam support and they just told me they added a command in the latest beta to re-import your categories from the old library to the new one, in the process you lose any new categories you may have made in the new client however.

> 
> To re-import your old categories into the Beta do the following with the Steam Client open and running the Beta:
> 1. Bring up the run window by holding Windows key and the r key
> 2. Put the following command into the run window's input box: steam://resetcollections
> 3. You will get a warning message that any new collections you made will be lost while re-importing your old categories. Selecting OK to continue.

I don't know if you can run it several times but the one time I did run it it worked and I got all my depressurizer created categories back.

**Comment by caenissnow** [comment]
That worked for me!  Nice!  Thank you!

**Comment by matrixraven** [comment]
What about creating new ones? I mean, what happens if you run depressurizer now? Is it working with the new beta?

**Comment by thefunkygibbon** [comment]
brilliant - yes steam://resetcollections worked for me. 
weirdly though it seems the new beta doesn't support library sharing properly.    I no longer have <friends name> lists any more.  yet i have a bunch of games in my 'uncatagorised' list which i'm pretty sure I don't personally own. (just checked and clicking on most of them it says "from the library of <friends name>. 
anyone else seeing this?

**Comment by Zhuinden** [comment]
You guys are lucky, running `steam://resetcollections` moved all my games from `Uncategorized` to `All`, but it still doesn't use my actual categories.

Funnily enough, the categories are still available in Big Picture Mode. :thinking:

**Comment by caenissnow** [comment]
@matrixraven  - I can confirm.  If you run Depressurizer again (I got several bundles this past week thanks to Fanatical's bundlefest) you do have to run steam://resetcollections again.  But it does work.  At least for me.

**Comment by wynick27** [comment]
@healingbrew I can confirm those collection ids are random numbers, generated by following code in libraryroot.js:
```
n.NewCollection = function(e, t, i) {
                var r = new n;
                return r.m_strId = function() {
                    var e = new Uint8Array(9);
                    return window.crypto.getRandomValues(e),
                    "uc-" + k.fromByteArray(e)
                }(),
                r.m_strName = e,
                r.m_filter = t,
                r.m_rgAddedManually = i,
                r.m_rgRemovedManually = [],
                r
            }
```
Actually, you can start steam client with -dev and press F12 to open developer tools and debug steam library. Some of the functions are registered in SteamUI.dll.

**Comment by neptuwunium** [comment]
I have successfully created custom steam categories programmatically in my own tag playground software, https://github.com/healingbrew/SteamCategorizer/commit/f9eef886910533414b75865d0d255c05e2818fed (This project was purely to toy with steam categories.) I strongly advise against using this program because it's unstable and a tendency to brick your collections/categories.

This is a generated category by that program (it contains all steam games with new library assets.)

![20190929155939](https://user-images.githubusercontent.com/614231/65834413-6ced2080-e2d2-11e9-8988-5eca79d4514b.png)

The logic is in the [FindGames() and SaveGames() methods in MainWindow.xaml](https://github.com/healingbrew/SteamCategorizer/blob/master/SteamCategorizer/MainWindow.xaml.cs#L89-L188)

Hopefully someone can make the necessary changes for Depressurizer to work.

Some things to note: 

1. Steam doesn't dispose CEF correctly, so local storage is almost always corrupt. The solution is to delete the Local Storage folder in appdata and relaunch Steam. This causes Steam to rebuild the data from what it uploaded to the cloud. Exiting at this point almost always has a clean leveldb storage.
2. Steam must NOT BE RUNNING while you access LevelDB, the database mutates easily and has a lockfile that will prevent reading if it's engaged. If for whatever reason you access it before Steam does, Steam will wipe local storage.
3. It is **NOT** Big Endian UTF-16, it's Little Endian UTF-16 prefixed with a null byte.
4. I don't know if it's actually UTF-16 for everyone, or only for some that have UTF-16 exclusive characters somewhere.

**Comment by TheMerricat** [comment]
One thing to be aware of is there seems to be a size limitation when importing your existing categories into the beta. If switching to the beta and/or using the resetcollections URL doesn't fix things for you, Try removing a portion of your categories via Depressurizer and reimporting.

**Comment by matrixraven** [comment]
I ended up going back to the non-beta because my categories work there. It seems my success was only the first time, and further refinements to my categorization doesn't carry over, even with the hints in this thread.

**Comment by Zhuinden** [comment]
Steam Support told me to use `steam://resetcollections`, I told them it doesn't work, they said "if you have no categories then unfortunately it won't be imported" despite sending them screenshots showing my categories, and it's been radio silence for a week now.

I guess the new library fails to import old categories *sometimes* and they never bothered to figure out why.

**Comment by matrixraven** [comment]
Well, it IS beta, so hopefully it'll be fixed before it goes into full release.

**Comment by Zhuinden** [comment]
Actually, I tested it today and now it seems to be working. 😁 I guess they looked into it and whatever bug was in the way has been fixed in some beta update :+1:

**Comment by matrixraven** [comment]
> 
> 
> Actually, I tested it today and now it seems to be working. 😁 I guess they looked into it and whatever bug was in the way has been fixed in some beta update 👍

Right you are! I just tried it, and finally got the popup that asked, "Are you sure you want to reimport your collections?" Clicking yes brought them all over. Thanks!

**Comment by Zhuinden** [comment]
I actually had some more obscure bug going on with my library (after doing the reset collections TWICE, it would never work again) and been mailing with Steam Support about it for a month, but now with the newest beta, my collections have actually been properly imported. :heart: 

Steam actually put in a lot of effort to fix it, I really like how much they cared about making it work.

**Comment by klepp0906** [comment]
> You guys are lucky, running `steam://resetcollections` moved all my games from `Uncategorized` to `All`, but it still doesn't use my actual categories.
> 
> Funnily enough, the categories are still available in Big Picture Mode. 🤔

same with me.  it actually creates all the categories, and initially shows them populated with the shortcuts, then crashes and after a restart, the categories all remain - but theyre unpopulated.

everything remains in BPM.

ultimately i rolled back to the old client and thats where ill stay until something gives with this hot mess.

(not to mention the new client is using 30% of a 8700k and 4gb of ram with the number of shortcuts i have) lol.

**Comment by Zhuinden** [comment]
@klepp0906 that's quite a surprise, the latest Beta version (and subsequently, most likely the actual Release version) *should* properly support things now.

I was having a chat with Steam Support and while it took a month, they identified a "bug that occurs quite rarely, but affects my library's category import" and the last beta successfully ran `steam://resetcollections`. Are you saying that even with the latest beta, your categories are **still** not imported?

Consider writing to Steam Support, they helped me resolve the issue. I sent them a bunch of screenshots and everything, and they fixed it based on that report.

**Comment by klepp0906** [comment]
Yes.  I just tried last night. 

Probably worth mentioning I have shortcuts which were created with SRM. I’m unsure if that has a hand in the problem or not. 

Even those that weren’t aren’t importing though. 

With some finagling I can get them to using resetcollections but it displays an import failed message and after restarting steam they’re back outside of the categories again. 

To add insult to injury, each time I’ve messed with it, somehow my smart sync with Dropbox breaks and I have to reinstall Dropbox. 

NO clue how that would Be even related, yet I’ve been running Dropbox for years and each time I upgrade to the new library and start trying to get it into a workable state, it ends up killing Dropbox again. Go figure. 

You might be right though, I should reach out to valve, perhaps it will bear fruit and save someone else from the trouble in the future. 

At current Im using the old client.

**Comment by Zhuinden** [comment]
Definitely write to Steam Support. They helped me too. I sent them a message in under "big picture mode" and it still worked ;)
